pkg/controller/run/parse_line.go

```diff
@@ -214,48 +211,52 @@ func (c *Controller) parseSemverTagLine(ctx context.Context, logE *logrus.Entry,
 			return patchLine(action, sha, lv), nil
 		}
 	}
-	// verify commit hash
-	if !c.param.IsVerify {
-		return "", nil
-	}
-	// @xxx # v3.0.0
-	// @<full commit hash> # v3.0.0
-	if FullCommitSHA != getVersionType(action.Version) {
-		return "", nil
-	}
-	if err := c.verify(ctx, action); err != nil {
-		return "", fmt.Errorf("verify the version annotation: %w", err)
-	}
-	return "", nil
-}
 
-func (c *Controller) parseShortSemverTagLine(ctx context.Context, logE *logrus.Entry, action *Action) (string, error) {
-	// @xxx # v3
-	// @<full commit hash> # v3
+	// ensure the commit sha is valid
 	if FullCommitSHA != getVersionType(action.Version) {
 		return "", ErrCantPinned
 	}
-	if c.param.Update {
-		lv, err := c.getLatestVersion(ctx, logE, action.RepoOwner, action.RepoName)
+
+	// versionParts is a slice of strings, each containing a group of
+	// consecutive numbers in the version comment, for example:
+	//     3.0.0   => []string{"3","0","0"}
+	//     1.23.44 => []string{"1","23","44"}
+	versionParts := regexp.MustCompile(`\d+`).FindAllString(action.VersionComment, -1)
+
+	// the number of parts that comprise a semantic version (MAJOR.MINOR.PATCH)
+	// ... this exists to avoid the error generated by go-mnd when using the int
+	// directly in the comparison below
+	semanticVersionParts := 3
+
+	// if we have a "short version" (e.g. not a complete `1.2.3` semver), check
+	// for a tag that matches the commit hash. note that this could be the same
+	// tag that's specified.
+	//
+	// using != supports validation against version tags like:
+	// - v1 (1 part)
+	// - v1.2 (2 parts)
+	// - v1.2.3.4 (4 parts)
+	if len(versionParts) != semanticVersionParts {
+		tag, err := c.getTagFromSHA(ctx, action, action.Version)
```

Comment:
pinact didn't change tags like `v1.2.3.4`, but this pull request tries to change them.
And it always gets tags by API, causing inefficient API calls.
We provide `--verify` option to verify sha, but without the option pinact should refrain inefficient API calls.

Reply:
It's good to convert `vX` and `vX.Y` to `vX.Y.Z`.
So pinact does it.
But we doesn't intend to convert `vA.B.C.D` to `vA.B.C.D.E`.
This is an edge case and in most cases this calls GitHub API uselessly.

Reply:
I don't think versions like `v1.2.3.4.5` are general.
Perhaps some actions have this kind of tags and update tags like `v1.2.3.4` when creating a new `v1.2.3.4.x`, but I think we don't need to consider them for now.

Reply:
agreed, i don't think i've seen someone do versions like that for a long time. i'll refactor this to only consider `< 3` parts in this block.

Reply:
also, i should note that the way the regex is (and was) written, it will always match `1.2.3.4.5...`.

---

> And it always gets tags by API, causing inefficient API calls.
> We provide --verify option to verify sha, but without the option pinact should refrain inefficient API calls.

Verification was never performed in `parseShortSemverTagLine()`, and that's what happens here. To be explicit, the behavior of performing the call for 4-parts-and-more isn't new: a version string that had 4 or more parts would have invoked `parseShortSemverTagLine()`, which performed a call. This little `if` check is effectively a replacement for `parseShortSemverTagLine()`.

Reply:
> Verification was never performed in parseShortSemverTagLine()

Ah, I see. Probably this is a bug.
Verification should be performed like `parseSemverTagLine`.
https://github.com/suzuki-shunsuke/pinact/blob/acde43cfcca49e7423cfa414da5c756337935821/pkg/controller/run/parse_line.go#L217-L220
https://github.com/suzuki-shunsuke/pinact/blob/acde43cfcca49e7423cfa414da5c756337935821/pkg/controller/run/parse_line.go#L226-L228

Reply:
You've linked to where verification is performed in `parseSemverTagLine`, not `parseShortSemverTagLine`.

Reply:
The changes made in this tree can be summed up as:

- `parseShortSemverTagLine` was merged into `parseSemverTagLine`, and renamed to `parseTagLine`.
- symbols referring to "short" or "long" tags were named to remove that distinction

Reply:
> You've linked to where verification is performed in parseSemverTagLine, not parseShortSemverTagLine.

This was intentional.
I linked the code of parseSemverTagLine as reference.

As I mentioned at https://github.com/suzuki-shunsuke/pinact/pull/942#discussion_r2072314207 ,
pinact distinguished short tags with semver, so I'm wondering if we should merge those functions.

Reply:
I saw it as unnecessary code that introduced complexity.

The `parseTagLine` function:

- deals with full semver tags in the same exact way as `parseSemverTagLine` did before
- deals with non-fully-semver "version tags" (aka "short semver tags") in the same way `parseShortSemverTagLine` did before